backend/lucene/src/main/java/org/hibernate/search/backend/lucene/search/query/impl/LuceneSearchResultExtractorImpl.java

```diff
@@ -88,23 +101,57 @@
 		return extractedData;
 	}
 
-	private void addChildrenContributions(IndexSearcher indexSearcher) throws IOException {
-		Document document = storedFieldVisitor.getDocument();
-		String id = document.getField( LuceneFields.idFieldName() ).stringValue();
-		ScoreDoc[] children = findChildren( indexSearcher, id );
-		for ( ScoreDoc childHit : children ) {
-			indexSearcher.doc( childHit.doc , storedFieldVisitor );
+	private Map<Integer, Set<Integer>> fetchNestedDocs(IndexSearcher indexSearcher, ScoreDoc[] scoreDocs) throws IOException {
+		// if the projection does not need any nested object skip their fetching
+		if ( !rootProjection.involveNestedFields() ) {
+			return new HashMap<>();
+		}
+
+		Map<String, Integer> parentIds = new HashMap<>();
+		for ( ScoreDoc hit : scoreDocs ) {
+			Document doc = indexSearcher.doc( hit.doc, ID_FIELD_SET );
+			String parentId = doc.getField( LuceneFields.idFieldName() ).stringValue();
+			if ( parentId == null ) {
+				continue;
+			}
+			parentIds.put( parentId, hit.doc );
+		}
+
+		Map<String, Set<Integer>> stringSetMap = fetchChildren( indexSearcher, parentIds.keySet() );
+		HashMap<Integer, Set<Integer>> result = new HashMap<>();
+		for ( Map.Entry<String, Set<Integer>> entry : stringSetMap.entrySet() ) {
+			result.put( parentIds.get( entry.getKey() ), entry.getValue() );
 		}
+		return result;
 	}
 
-	public ScoreDoc[] findChildren(IndexSearcher indexSearcher, String id) {
-		TermQuery query = new TermQuery( new Term( LuceneFields.rootIdFieldName(), id ) );
+	private Map<String, Set<Integer>> fetchChildren(IndexSearcher indexSearcher, Set<String> parentIds) {
+		BooleanQuery.Builder builder = new BooleanQuery.Builder();
+		for ( String parentId : parentIds ) {
+			TermQuery query = new TermQuery( new Term( LuceneFields.rootIdFieldName(), parentId ) );
+			builder.add( query, BooleanClause.Occur.SHOULD );
+		}
+		builder.setMinimumNumberShouldMatch( 1 );
+
+		BooleanQuery booleanQuery = builder.build();
```

Comment:
If you're going to use a query, to fetch children, you'd better use `ToChildBlockJoinQuery`, which might be more optimized. This seems to be an analog to what we already use for `nested` predicates.

We could do something like this:

```
new ToChildBlockJoinQuery(
    new MatchAllDocsQuery(),
    new QueryBitSetProducer( luceneQuery ),
    ScoreMode.None
)
```

Where `luceneQuery` is the query that was used to retrieve the parents.

As you can see, it requires to pass a `BitSetProducer` to specify the parents we're interested in, and the only available implementation is based on a query (`QueryBitSetProducer`), so we use that.
But in our case, we just executed a query to retrieve the parents, so it's a bit ridiculous.

We should think about it... I'm sure there's a better way. Maybe we could turn the `TopDocs` into a `BitSet`, or collect a `DocIdSet` in `org.hibernate.search.backend.lucene.search.extraction.impl.LuceneCollectors#collect` when we know we will need to load data from nested documents.

Reply:
cool